### PR TITLE
Added hseqr! LAPACK function

### LINKED ISSUE #962: Add LAPACK functions for Nonsymmetric Eigenvalues Problems
Hello,

I think it could be great to add some of the LAPACK functions for [Nonsymmetric Eigenvalues Problems](https://www.intel.com/content/www/us/en/develop/documentation/onemkl-developer-reference-fortran/top/lapack-routines/lapack-least-squares-and-eigenvalue-problem/lapack-least-square-eigenvalue-problem-computation/nonsymm-eigenvalue-problems-lapack-computation.html). 

Most of them [are yet implemented](https://docs.julialang.org/en/v1/stdlib/LinearAlgebra/#LinearAlgebra.LAPACK.gehrd!), but I noticed that for example `hseqr` (which finds the eigenvalues and Schur factorization of an Hesenber matrix) is not yet present.

### LINKED ISSUE #47291: Add LAPACK functions for Nonsymmetric Eigenvalues Problems
Hello,

I think it could be great to add some of the LAPACK functions for [Nonsymmetric Eigenvalues Problems](https://www.intel.com/content/www/us/en/develop/documentation/onemkl-developer-reference-fortran/top/lapack-routines/lapack-least-squares-and-eigenvalue-problem/lapack-least-square-eigenvalue-problem-computation/nonsymm-eigenvalue-problems-lapack-computation.html). 

Most of them [are yet implemented](https://docs.julialang.org/en/v1/stdlib/LinearAlgebra/#LinearAlgebra.LAPACK.gehrd!), but I noticed that for example `hseqr` (which finds the eigenvalues and Schur factorization of an Hesenber matrix) is not yet present.

### PR DESCRIPTION
**Pull request by albertomercurio**
This fixes JuliaLang/LinearAlgebra.jl#962

### CODE REVIEW

**Comment by albertomercurio** [comment]
Ok, I think that's all.

**Comment by dkarrasch** [comment]
Nice. Is that copy-paste from @Jutho's package? We definitely need tests for this. Also, could/should we add dispatch rules for the `eigen` function to call this whenever possible and meaningful? Like `norm` calls BLAS for strided arrays with the usual floating point numbers etc.?

File: stdlib/LinearAlgebra/src/lapack.jl [hunk 1]
```diff
@@ -5741,6 +5741,104 @@ for (ormhr, elty) in
...
+                    lwork = BlasInt(real(work[1]))
+                    resize!(work, lwork)
+                end
+            end
+            H, Z, complex.(wr, wi)
+        end
+    end
+end
+hseqr!(H::AbstractMatrix, Z::AbstractMatrix) = hseqr!('S', 'V', 1, size(H, 1), H, Z)
+hseqr!(H::AbstractMatrix) = hseqr!('S', 'V', 1, size(H, 1), H, one(H))
```
Comment (dkarrasch) [inline stdlib/LinearAlgebra/src/lapack.jl:5829]: Perhaps restrict here to the only (el)types that work (at least I assume that)?

```suggestion
hseqr!(H::StridedMatrix{T}, Z::AbstractMatrix{T}) where {T<:BlasFloat} = hseqr!('S', 'V', 1, size(H, 1), H, Z)
hseqr!(H::StridedMatrix{T}) where {T<:BlasFloat} = hseqr!('S', 'V', 1, size(H, 1), H, one(H))
```

**Comment by albertomercurio** [comment]
> Nice. Is that copy-paste from @Jutho's package?

Yes, but with a bit of improvement.

> Also, could/should we add dispatch rules for the `eigen` function to call this whenever possible and meaningful?

The `hseqr!` function computes the eigenvalues and, optionally, the Schur decomposition. It never gives you the eigenvectors. Thus, I think the correct comparison should be with the `schur` function, which calculates the factorization for a general matrix.

I will do it soon.

**Comment by dkarrasch** [comment]
One last gentle request: can you test that the promotion works as desired? Like apply your tests to a small integer `UpperHessenberg` matrix.

**Review by dkarrasch — APPROVED** [review]

**Comment by Jutho** [comment]
I have been following and I am very happy to see this. That takes away a burden of the maintenance of KrylovKit.

**Comment by albertomercurio** [comment]
Ho, I think that's all.

**Pull request merged**